### ISSUE TITLE
global.order function breaks syntax highlighting

### ISSUE THREAD

**Issue by patriciogonzalezvivo** [tangrams/tangram-play#449]
![screen shot 2016-07-19 at 10 48 09 am](https://cloud.githubusercontent.com/assets/346914/16954302/629292cc-4d9e-11e6-8f5c-bd5fa99cb81d.png)

**Comment by louh** [comment]
@patriciogonzalezvivo, I've changed the wording - assuming you mean "syntax highlighting" and not other kinds of highlighting (like line highlighting)?

**Comment by louh** [comment]
Can you paste in your YAML file here? Do you have any idea why `global.order` of line 49 breaks syntax highlighting, but not the `global.order` of line 42? I would like to look into this further because it appears something else might be going on, not `global.order` per se.

**Comment by patriciogonzalezvivo** [comment]
yep... from that line to the end all syntax highlighting stop working properly

**Comment by patriciogonzalezvivo** [comment]
```yaml
sources:
    mapzen:
        type: TopoJSON
        url: https://vector.mapzen.com/osm/all/{z}/{x}/{y}.topojson?api_key=vector-tiles-P6dkVl4

cameras:
    perspective:
        type: flat

global:
    order_add_9: function () { return (feature.sort_key * 20 + 9); }
    order_add_8: function () { return (feature.sort_key * 20 + 8); }
    order_add_7: function () { return (feature.sort_key * 20 + 7); }
    order_add_6: function () { return (feature.sort_key * 20 + 6); }
    order_add_5: function () { return (feature.sort_key * 20 + 5); }
    order_add_4: function () { return (feature.sort_key * 20 + 4); }
    order_add_3: function () { return (feature.sort_key * 20 + 3); }
    order_add_2: function () { return (feature.sort_key * 20 + 2); }
    order_add_1: function () { return (feature.sort_key * 20 + 1); }
    order: function () { return (feature.sort_key * 20); }
    order_sub_1: function () { return (feature.sort_key * 20 - 1); }
    order_sub_2: function () { return (feature.sort_key * 20 - 2); }
    order_sub_3: function () { return (feature.sort_key * 20 - 3); }
    order_sub_4: function () { return (feature.sort_key * 20 - 4); }
    order_sub_5: function () { return (feature.sort_key * 20 - 5); }
    order_sub_6: function () { return (feature.sort_key * 20 - 6); }
    order_sub_7: function () { return (feature.sort_key * 20 - 7); }
    order_sub_8: function () { return (feature.sort_key * 20 - 8); }
    order_sub_9: function () { return (feature.sort_key * 20 - 9); }

layers:
    earth:
        data: { source: mapzen }
        draw:
            polygons:
                order: global.order
                color: '#555'
    water:
        data: { source: mapzen }
        draw:
            polygons:
                order: global.order
                color: '#333'

    landuse:
        data: { source: mapzen }
        draw:
            polygons:
                order: global.order
                color: '#666'
    roads:
        data: { source: mapzen }
        draw:
            lines:
                order: global.order
                color: '#FFF'
                width: 1px
    buildings:
        data: { source: mapzen }
        draw:
            polygons:
                order: global.order
                color: '#999'


```

**Comment by louh** [comment]
This is occurring because of (1) how we're checking whether some value is a JavaScript value, and (2) how Tangram compiles `scene.config`.

In Tangram Play, we switch the CodeMirror's mode to JavaScript when we detect that the value of that line will be a JavaScript value. It does this by looking at the Tangram scene config for that address and checking if the value begins with `function () {`.

In Tangram, the scene config is compiled. In other words, if you look at `scene.config.water.draw.polygon`, you don't get back a string containing `global.order`, which is what you see in the editor. You get back a string containing a JavaScript function.

Therefore, CodeMirror's parser looks in Tangram, sees that the value of `global.order` _looks like_ a JavaScript function, and starts highlighting it accordingly. Once it's in JavaScript mode, it's not properly falling back out of it, either.

The best fix for the moment will be to parse & switch modes based on the content _in the editor_ and not by checking against the compiled Tangram scene config object, which no longer matches what you see in the editor itself.

**Comment by louh** [comment]
@patriciogonzalezvivo, please review this fix:

PR: https://github.com/tangrams/tangram-play/pull/454
Precog: https://precog.mapzen.com/tangrams/tangram-play/inner-mode-fix/

Before:


![screenshot 2016-07-19 19 49 16](https://cloud.githubusercontent.com/assets/2553268/16970430/eee843c0-4de9-11e6-8ce7-724ea81017bb.png)

After:

![screenshot 2016-07-19 19 49 00](https://cloud.githubusercontent.com/assets/2553268/16970428/ed02330e-4de9-11e6-9cd2-a06a79d3a8de.png)

**Comment by louh** [comment]
Closed via #454.
